### ISSUE TITLE
Connection failure and SSL handshake error when using libh2o as a http1client

### ISSUE THREAD

**Issue by shazni** [h2o/h2o#2441]
First of all, to mention, h2o is superb. Great stuff. I've been trying to use it develop a server for my application and also to use as a client using the libh2o. 

I'm having a few issues dealing in using h2o as a client. I've read the http1client.c in the example directory and also have looked at the proxy.c. I'm writing a C++ class that can act as an http client. I'm wrapping the h2o constructs within it. I have had some success and few issues which I couldn't figure out why. 

First issue is I get 'ssl handshake error' when trying to fetch a https page. This issue is reproducible in the http1client.c in the 2.2.6 version.

Second issue is I use the client against a server I have implemented using h2o. Server works fine and has few paths registered; say post-test, get-test. I use the C++ client (that wraps h2o constructs) to connect to the server and I see 'connection failures' to POST requests where as some prevuous POST requests for the same resource succeeds. I wonder if I do anything incorrectly in my client code. I use another C++ http client to connect to the same and it always passes. Here is a snippet of my HttpClient. Appreciate if you could review my code to see anthing unusual or incorrect. I have taken some code from http1client.c as well as from the proxy.c file.

`H2OHttpResponse* H2OHttpClient::call(const char* method) {    
    if (h2o_url_parse(url.c_str(), SIZE_MAX, &url_parsed) != 0) {
        boost::format fmt = boost::format("Unrecognized type of URL: %s") % url;
        throw HydrogenH2OException(fmt.str());
    }

    is_ssl = url_parsed.scheme == &H2O_URL_SCHEME_HTTPS;

    if(is_ssl) {
        SSL_load_error_strings();
        SSL_library_init();
        OpenSSL_add_all_algorithms();
        ctx.ssl_ctx = SSL_CTX_new(TLSv1_client_method());
        if(certificatePath != "") {
            // TODO - check if the certificate path actually exists
            SSL_CTX_load_verify_locations(ctx.ssl_ctx, certificatePath.c_str(), NULL);
        }
        if(peerVerification) {
            SSL_CTX_set_verify(ctx.ssl_ctx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
        } else {
            SSL_CTX_set_verify(ctx.ssl_ctx, SSL_VERIFY_NONE, NULL);
        }
    }

    h2o_mem_init_pool(&pool);

#if H2O_USE_LIBUV
    ctx.loop = uv_loop_new();
#else
    ctx.loop = h2o_evloop_create();
#endif

    h2o_multithread_queue_t *queue = h2o_multithread_create_queue(ctx.loop);
    h2o_multithread_register_receiver(queue, ctx.getaddr_receiver, h2o_hostinfo_getaddr_receiver);
    h2o_timeout_init(ctx.loop, &io_timeout, timeoutMs);

    setClientActive(true);

    h2o_mem_clear_pool(&pool);

    self = (client_data *) h2o_mem_alloc_shared(&pool, sizeof(*self), HydrogenH2OHttpClient::on_generator_dispose);
    self->data = this;

    h2o_iovec_t buf;
    size_t offset = 0;
    buf.len = strlen(method) + url_parsed.path.len + url_parsed.authority.len + 512;
    buf.base = (char *) h2o_mem_alloc_pool(&pool, buf.len);

    // Setting up buf to include the headers and body
    // ... 
    
    assert(offset <= buf.len);
    buf.len = offset;

    self->req.bufs[0] = buf;
    std::string headerBuffer(buf.base, buf.len);
    logger.log(DEBUG, "[HydrogenH2OHttpClient] =====> header buffer = {}", headerBuffer);
    if(strcmp(method, "HEAD")) {
        self->req.is_head = 0;
    }

    h2o_iovec_t requestBuf;
    if(requestBody.size() > 0) {
        assert(strcmp(method, "GET"));
        requestBuf.len = requestBody.size() + 1;
        requestBuf.base = (char *) h2o_mem_alloc_pool(&pool, requestBody.size() + 1);
        memcpy(requestBuf.base, requestBody.c_str(), requestBody.size());
    } else {
        requestBuf.len = 0;
        requestBuf.base = NULL;
    }
    self->req.bufs[1] = requestBuf;
    
    /// end setting up ---

    assert(offset <= buf.len);
    buf.len = offset;

    self->req.bufs[0] = buf;

    h2o_iovec_t requestBuf;
    if(requestBody.size() > 0) {
        assert(strcmp(method, "GET"));
        requestBuf.len = requestBody.size() + 1;
        requestBuf.base = (char *) h2o_mem_alloc_pool(&pool, requestBody.size() + 1);
        memcpy(requestBuf.base, requestBody.c_str(), requestBody.size());
    } else {
        requestBuf.len = 0;
        requestBuf.base = NULL;
    }
    self->req.bufs[1] = requestBuf;
    
    clientConnect();

    return &h2oHttpResponse;
}

void H2OHttpClient::clientConnect() {
    /// ### Is this correct? Should I destroy the socketpool for every request?
    if (sockpool != NULL) {
        // If socketpool hasn't been destroyed. Only a safety check
        h2o_socketpool_dispose(sockpool);
        free(sockpool);
        sockpool = NULL;
    }

    if (1) {
        if (sockpool == NULL) {
            sockpool = (h2o_socketpool_t *) h2o_mem_alloc(sizeof(*sockpool));
        }
        h2o_socketpool_init_by_hostport(sockpool, url_parsed.host, h2o_url_get_port(&url_parsed), is_ssl, 10);
        h2o_socketpool_set_timeout(sockpool, ctx.loop, 5000); // Timeout in milliseconds
        h2o_http1client_connect_with_pool(NULL, self, &ctx, sockpool, H2OHttpClient::on_connect);
    } else {
        h2o_http1client_connect(NULL, self, &ctx, url_parsed.host, h2o_url_get_port(&url_parsed), is_ssl, HydrogenH2OHttpClient::on_connect);
    }

    while (isClientActive) {
        printf("Running the event loop in Client\n");
#if H2O_USE_LIBUV
        uv_run(ctx.loop, UV_RUN_ONCE);
#else
        h2o_evloop_run(ctx.loop, INT32_MAX);
#endif
    }
}`

I have couple of questions
1) socketpool is a static variable. Is it thread safe? If I use the client in a multi-threaded application, will it fail?
2) All this is in a C++ class. Does it make sense to call h2o_socketpool_dispose() after each class instance goes out of scope? Or does it defeat the purpose of using a sockepool in the first place?

I also noticed that beta2 of the 3.0.0 version is out and it has http1client and htt2client. It would be great to have some example usage of that in libh2o point of view.

**Comment by shazni** [comment]
I have made some progress here. I used the the 2.3.0-beta2 and changed the client side code. Learnt fair amount of details looking at src/httpclient.c (which seems to be moved from the examples directory) and the proxy.c. 

SSL issue is solved by using TLSv1_2_client_method() instead of TLSv1_client_method(). 

Then I had this connection failure with no clue. This apparently seems to be linked with Ipv6. Confusing part is, some of my test cases that use 'localhost' to connect to a locally running server succeed and some failed with a 'connection failure'. However, when I used loopback IP address it had no trouble with 'connection failure'. Also when I switched off Ipv6 (I'm using a Fedora 32) even using 'localhost' had no trouble. 

Is there a way to force h2o to only consider Ipv4 resolution from the client side?

Also with 2.3.0-beta2 I have near success except now I face 'refused stream' error in on_body with HTTP/2 400 when the content-length header is set. When it's not set for a POST request it doesn't throw 'refused stream' error, however my response from an echo server doesn't seem to be properly null terminated. So I think I need to set the content-length properly. 

I use the below snippet to add that header. 

```
        h2o_headers_t headers_vec = (h2o_headers_t) {NULL};
        std::string _requestBody = "Some payload string here";
        const char* requestBody = _requestBody.c_str();
        h2o_iovec_t cl_buf = build_content_length(self->pool, strlen(requestBody));
        h2o_add_header(self->pool, &headers_vec, H2O_TOKEN_CONTENT_LENGTH, NULL, cl_buf.base, cl_buf.len);  

        *headers = headers_vec.entries;
        *num_headers += 1;
```

I used this from the src/httpclient.c.

Is this method correct? 

Also in src/httpclient.c the chunk_size is set to 10 and in that scenario should the content-length be only set in the on_connect to full length of the request only once or should it be somehow separated in the fill_body function to set each time when we send a chunk?
